articles/introduction-to-theory-of-computation/index.md

```diff
@@ -0,0 +1,138 @@
+
+[Automata Theory](https://en.wikipedia.org/wiki/Automata_theory) is the study of abstract computational devices. It forms a formal framework for designing and analyzing computing devices such as biocomputers and quantum computers. These models are essential in several areas of computation (applied and theoretical fields). 
+
+An Automaton is a machine that operates singly on input and follows a defined pattern or configuration to produce the desired output. Through automata, we learn how problems and compute functions are solved by the use of automatons. 
+
+A basic computation performed on/by an automaton is defined by the following features:
+
+- A set of input symbols.
+
+- The configuration states.
+
+- Output.
+
+#### Branches of Automata Theory 
+- **Finite Automata(FA):** This is a computer model that is inferior in its computation ability. This model is fit for devices with limited memory. It is a simple abstract machine with five elements that define its functioning and processing of problems.
+
+A [*Finite Automaton* (FA)](https://www.geeksforgeeks.org/introduction-of-finite-automata/) is a finite collection of states with rules (transition functions) for traversing through the states depending on the input symbol. FA accepts or rejects input strings while reading the strings from left to right.
+
+**The tuples are:**
+
++ **Q:** *Finite set of states*
+
++ **∑:** *Set of input symbols*
+
++ **q:** *Initial state*
+
++ **F:** *Set of final states*
+
++ **δ:** *Transition function*
+
+Finite Automata is useful in building text editors/text preprocessors. FA are poor models of computers. They can only perform simple computational tasks.
+
+- **Context-Free Grammars (CFGs):**  They are more powerful abstract models than FA and are essentially used in the programming languages and natural language research work.
+
+- **Turing Machines:** They are abstract models for real computers having an infinite memory (in the form of a tape) and a reading head. They form powerful computation models than FA, CFGs, and Regular Expressions.
+
+### Computability Theory
+The [Computability theory](https://plato.stanford.edu/entries/computability/) defines whether a problem is “solvable” by any abstract machine. Some problems are computable while  others are not.
+
+Computation is done by various computation models depending on the nature of the problem at hand, among these machines being; Turing machine, Finite state machines, and many others.
+
+### Complexity Theory
+This branch is about studying the cost of solving problems while focusing on resources (time & space) needed as the metric.
+
+**Measuring Complexity:**
+
+Involves analysis of the algorithm to determine how much time it takes while solving a problem. As T(n), running time is usually a complex expression when an algorithm runs on large input sizes. This would mean an exponential growth in the running time hence cost.
```

Comment:
@OMONDIELLY 

What does T(n) mean here? And why is the running time usually a complex expression when the input size is large?
Isn't there a possibility of algorithms running on O(n) or O(1)?

I think this line is a bit unclear. Can you explain this a bit better?